rust/routee-compass-core/src/model/cost/cost_model.rs

```diff
@@ -0,0 +1,234 @@
+            self.network_state_variable_rates.clone(),
+        )?;
+        let network_cost = self.cost_aggregation.agg(&network_costs);
+        let total_cost = vehicle_cost + network_cost;
+        let pos_cost = Cost::enforce_strictly_positive(total_cost);
+        Ok(pos_cost)
+    }
+
+    /// Calculates a cost estimate for traversing between a source and destination
+    /// vertex without actually doing the work of traversing the edges.
+    /// This estimate is used in search algorithms such as a-star algorithm, where
+    /// the estimate is used to inform search order.
+    ///
+    /// # Arguments
+    ///
+    /// * `src_state` - state at source vertex
+    /// * `dst_state` - estimated state at destination vertex
+    ///
+    /// # Returns
+    ///
+    /// Either a cost estimate or an error. cost estimates may be
+    pub fn cost_estimate(
+        &self,
+        src_state: &[StateVar],
+        dst_state: &[StateVar],
+    ) -> Result<Cost, CostError> {
+        let vehicle_costs = cost_ops::calculate_vehicle_costs(
+            src_state,
+            dst_state,
+            &self.state_variable_indices,
+            self.state_variable_coefficients.clone(),
+            self.vehicle_state_variable_rates.clone(),
+        )?;
+        let vehicle_cost = self.cost_aggregation.agg(&vehicle_costs);
+        let pos_cost = Cost::enforce_non_negative(vehicle_cost);
+        Ok(pos_cost)
+    }
+
+    /// Serializes the cost of a traversal state into a JSON value.
+    ///
+    /// # Arguments
+    ///
+    /// * `state` - the state to serialize
+    ///
+    /// # Returns
+    ///
+    /// A JSON serialized version of the state. This does not need to include
+    /// additional details such as the units (kph, hours, etc), which can be
+    /// summarized in the serialize_state_info method.
+    fn serialize_cost(&self, state: &TraversalState) -> Result<serde_json::Value, CostError> {
+        let mut state_variable_costs = self
+            .state_variable_indices
+            .iter()
+            .filter(|(name, _)| self.vehicle_state_variable_rates.contains_key(name))
+            .map(move |(name, idx)| {
+                let state_var = state
+                    .get(*idx)
+                    .ok_or(CostError::StateIndexOutOfBounds(*idx, name.clone()))?;
+                let rate = self
+                    .vehicle_state_variable_rates
+                    .get(name)
+                    .ok_or(CostError::StateVariableNotFound(name.clone()))?;
+                let cost = rate.map_value(*state_var);
+                Ok((name.clone(), cost))
+            })
+            .collect::<Result<HashMap<String, Cost>, CostError>>()?;
+
+        let total_cost = state_variable_costs
+            .values()
+            .fold(Cost::ZERO, |a, b| a + *b);
+        state_variable_costs.insert(String::from("total_cost"), total_cost);
+
+        let result = serde_json::json!(state_variable_costs);
```

Comment:
Would it be possible to add an additional field here, something like `total_realized_cost` that captures the sum of the cost that the model actually optimized for since total cost will be the same regardless of the coefficients that are used.

Reply:
that appears as the "cost" field in the base level of the reponse, which gets placed there by the summary output plugin. does that cover your ask?

Reply:
to disambiguate, i just renamed that "route_cost".

also, looking at the summary output plugin, i'm thinking all of this functionality could probably be moved into the CompassApp.apply_output_processing method, and we could remove the JSON summary extensions and summary plugin.

Reply:
> that appears as the "cost" field in the base level of the reponse, which gets placed there by the summary output plugin. does that cover your ask?

ohhh yeah I missed that

> I'm thinking all of this functionality could probably be moved into the CompassApp.apply_output_processing method, and we could remove the JSON summary extensions and summary plugin.

yeah agreed, that would make sense